src/Controller/Controller.php

```diff
@@ -223,6 +225,9 @@ public function __construct(
         if ($eventManager !== null) {
             $this->setEventManager($eventManager);
         }
+        if ($container !== null) {
+            $this->_components = new ComponentRegistry($this, $container);
+        }
```

Comment:
Having `container` passed into controller constructors could go awry. It will be the first time that we have made it easy for application code to get access to the container in a direct simple way. I don't think there is a backwards compatible way to build this right now if we don't want to pass the container in. I'm ok with moving this to 6.x and cleaning it up there too.

Reply:
> Having container passed into controller constructors could go awry.

Yup. I too had looked into adding DI support for components a while back and had come up with [this](https://github.com/cakephp/cakephp/compare/5.next...ADmad:cakephp:5.x-component-di). 

The difference is instead of passing the container instance to the controller I am passing in a component registry instance. It did create a bit of a chicken and egg problem between the controller and component registry which is why I had to add a method to set the controller instance to the registry later. So we can weigh the pros and cons of both approaches.

I didn't open a PR for my changes is because I question the value/need for components all together, with the availability of  traits and the ability to add middlewares for controllers. Only benefit I see is the convention based auto run event callbacks.

Reply:
> The difference is instead of passing the container instance to the controller I am passing in a component registry instance. It did create a bit of a chicken and egg problem between the controller and component registry which is why I had to add a method to set the controller instance to the registry later. So we can weigh the pros and cons of both approaches.

Passing ComponentRegistry to `Controller` might be the better solution, as it doesn't require changing controller's constructor.

> I didn't open a PR for my changes is because I question the value/need for components all together, with the availability of traits and the ability to add middlewares for controllers. 

Having components able to use services does give folks a way to incrementally refactor their application away from components and use services instead.

Reply:
> ...as it doesn't require changing controller's constructor.

My branch does the controller's constructor and add component registry as an argument.

> Having components able to use services does give folks a way to incrementally refactor their application away from components and use services instead.

Okay, I am not against this change then.